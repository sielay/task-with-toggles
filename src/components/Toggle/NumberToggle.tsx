import React, { FC } from "react";
import { Select } from "../Select";
import { Switch } from "../Switch";
import { ToggleProps } from "./types";

export const NumberToggle: FC<ToggleProps<number>> = ({
  value,
  onChange,
  options,
  disabled,
  label,
}) => {
  const booleanValue = !!(value && value > 0);
  const positiveValue = (options && options.filter(Boolean)[0]) || 0;
  return (
    <div className="Toggle">
      <div className="Toggle__label">{label}</div>
      <Select
        options={options && options.map((value) => `${value}`)}
        value={value !== undefined ? `${value}` : undefined}
        onChange={(value) => {
          onChange && onChange(parseInt(value));
        }}
      ></Select>
      <Switch
        value={booleanValue}
        onChange={(value) => {
          onChange && onChange(value ? positiveValue : 0);
        }}
        disabled={disabled}
      />
    </div>
  );
};
