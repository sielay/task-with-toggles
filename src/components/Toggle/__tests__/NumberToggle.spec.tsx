import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { fireEvent } from "@testing-library/dom";
import { NumberToggle } from "../NumberToggle";

describe("components.Toggle.NumberToggle", () => {
  it("Renders as expected", () => {
    const { baseElement } = render(
      <NumberToggle
        value={1}
        options={[1, 2, 3]}
        onChange={() => {
          // noop
        }}
        label="ABC"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
  it("Handle value changes", async () => {
    const onChange = jest.fn();
    render(
      <NumberToggle
        value={1}
        options={[1, 2, 3]}
        onChange={onChange}
        label="ABC"
      />
    );
    userEvent.click(screen.getByRole("checkbox"));
    await waitFor(() => expect(onChange).toBeCalledTimes(1));
    expect(onChange).toBeCalledWith(0);
    /* eslint-disable testing-library/no-node-access */
    const select = screen.getByText('2').parentNode;
    select && fireEvent.change(select, { target: { value: 2 } })
    await waitFor(() => expect(onChange).toBeCalledTimes(2));
    expect(onChange).toBeCalledWith(2);
  });
});
