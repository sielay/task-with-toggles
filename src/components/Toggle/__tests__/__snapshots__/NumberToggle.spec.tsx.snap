// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components.Toggle.NumberToggle Renders as expected 1`] = `
<body>
  <div>
    <div
      class="Toggle"
    >
      <div
        class="Toggle__label"
      >
        ABC
      </div>
      <select
        class="Select"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
      </select>
      <label
        aria-checked="true"
        class="Switch Switch--checked"
        role="checkbox"
        tabindex="0"
      >
        <input
          checked=""
          role="none"
          type="checkbox"
        />
        <span />
      </label>
    </div>
  </div>
</body>
`;
